### PR TITLE
Reimplement shell splitting

### PR DESCRIPTION
**Pull request by mattgodbolt**
Takes the bones of shell-split and makes it handle lots of situations without throwing exceptions, and using modern-ish typescript/javascript stuff.

Adds lots more tests too.

Removes unneeded package dependencies.

Changes from before: escaping a pipe character `moo \\# hello` would drop the `# hello` part (?) that is, I updated the test that did:
```ts
expect(splitArguments('hello \\#veryfancy etc')).toEqual(['hello', '\\']);
```
to be the more sane expectation of `['hello', '\\#veryfancy']`.

Everything else I could think of (unterminated strings, escapes etc) I kept the same. And added tests for.

Hopefully addresses #7201 without breaking any reasonable use of the code (though @partouf raises fair concerns about this.)

### CODE REVIEW

**Review by AbrilRBS — APPROVED** [review]

File: shared/common-utils.ts [hunk 1]
```diff
@@ -102,10 +102,110 @@ export function addDigitSeparator(n: string, digitSeparator: string, chunkSize:
     return splitIntoChunks(n, chunkSize).join(digitSeparator);
 }
 
-export function splitArguments(options = ''): string[] {
-    // escape hashes first, otherwise they're interpreted as comments
-    const escapedOptions = options.replace(/#/g, '\\#');
-    return quoteParse(escapedOptions)
-        .map((x: any) => (typeof x === 'string' ? x : (x.pattern as string)))
-        .filter(Boolean);
+class ArgSplitter {
```
Comment (junlarsen) [inline shared/common-utils.ts:105]: not very important but maybe we could call it something like ArgumentParser or something

I can also imagine myself changing this to a parser combinator or something in the future perhaps

**Review by junlarsen — APPROVED** [review]

File: shared/common-utils.ts (hunk 1 above)
Comment (mattgodbolt) [reply shared/common-utils.ts:105]: Will rename: and you'd be welcome to rephrase as a combinator etc, I was trying to make minimal(ish) changes from a mostly-known-good source (and add enough tests to feel happy about things!)

**Comment by mattgodbolt** [comment]
Ah shucks, forgot to push my rename. will push straight to main

**Comment by mattgodbolt** [comment]
This is rolling out to production now.

**Pull request merged**